segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DurableLog.java

```diff
@@ -147,59 +144,56 @@ protected void doStart() {
         long traceId = LoggerHelpers.traceEnterWithContext(log, traceObjectId, "doStart");
         log.info("{}: Starting.", this.traceObjectId);
 
-        this.executor.execute(() -> {
-            try {
-                boolean anyItemsRecovered = performRecovery();
-                this.operationProcessor.startAsync().awaitRunning();
-                if (!anyItemsRecovered) {
-                    // If the DurableLog is empty, need to queue a MetadataCheckpointOperation so we have a valid starting state (and wait for it).
-                    queueMetadataCheckpoint().get(RECOVERY_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
-                }
-            } catch (Exception ex) {
-                if (this.operationProcessor.isRunning()) {
-                    // Make sure we stop the operation processor if we started it.
-                    this.operationProcessor.stopAsync();
-                }
-
-                notifyFailed(ExceptionHelpers.getRealException(ex));
-                return;
-            }
+        // Initiate recovery.
+        CompletableFuture.supplyAsync(this::performRecovery, this.executor)
+                .thenCompose(anyItemsRecovered -> ServiceHelpers.startAsync(this.operationProcessor, this.executor)
+                        .thenComposeAsync(v -> anyItemsRecovered ? CompletableFuture.completedFuture(null) : queueMetadataCheckpoint(), this.executor))
```

Comment:
Ah but I do! You see how I make use of `anyItemsRecovered`? If I chain them I'd lose that one. The alternative would be to store that boolean into an `AtomicBoolean` and reuse it, which would make the code sloppier.